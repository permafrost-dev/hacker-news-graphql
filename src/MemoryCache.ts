export interface CacheItem {
    ts: number;
    expires: number;
    value: any;
}

export interface CacheMap {
    [key: string]: CacheItem;
}

export class MemoryCache {
    public map: CacheMap = {};

    public has(key: string): boolean {
        this.purge();

        return typeof this.map[key] !== 'undefined';
    }

    public get(key: string): any {
        this.purge();

        return this.map[key]?.value ?? null;
    }

    public put(key: string, value: any, ttlSeconds: number) {
        this.map[key] = {
            ts: new Date().getTime(),
            expires: new Date().getTime() + ttlSeconds * 1000,
            value: value,
        };
    }

    public purge() {
        for (const prop in this.map) {
            if (new Date().getTime() >= this.map[prop].expires) {
                delete this.map[prop];
                continue;
            }
        }
    }
}
