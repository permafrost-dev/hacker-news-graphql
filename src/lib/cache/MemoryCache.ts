import { Cache } from './Cache';

export interface CacheItem {
    ts: number;
    expires: number;
    value: any;
}

export interface CacheMap {
    [key: string]: CacheItem;
}

export class MemoryCache extends Cache {
    public map: CacheMap = {};

    constructor(keyPrefix = '') {
        super();

        if (keyPrefix.length) {
            keyPrefix += ':';
        }

        this.prefix = keyPrefix;
    }

    protected makeKey(key: string): string {
        return `${this.prefix}${key}`;
    }

    public async has(key: string): Promise<boolean> {
        this.purge();

        return new Promise(resolve => resolve(typeof this.map[this.makeKey(key)] !== 'undefined'));
    }

    public async get(key: string): Promise<any | null> {
        this.purge();

        const result = this.map[this.makeKey(key)]?.value ?? null;

        return new Promise(resolve => resolve(result));
    }

    public put(key: string, value: any, ttlSeconds: number) {
        this.map[this.makeKey(key)] = {
            ts: new Date().getTime(),
            expires: new Date().getTime() + ttlSeconds * 1000,
            value: value,
        };
    }

    public purge() {
        for (const prop in this.map) {
            if (new Date().getTime() >= this.map[prop].expires) {
                delete this.map[prop];
                continue;
            }
        }
    }

    public clear() {
        //
    }
}
