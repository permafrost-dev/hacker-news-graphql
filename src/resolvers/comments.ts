/* eslint-disable no-unused-vars */

import axios from 'axios';

export function getResolver(cache: any): any {
    const commentsResolver = async (parent, { first }) => {
        const comments: Record<string, any>[] = [];

        if (typeof parent['kids'] === 'undefined') {
            return new Promise(resolve => resolve([]));
        }

        const commentIds = parent.kids.slice(0, 1);

        for (const id of commentIds) {
            if (!cache.has(`comment:${id}`)) {
                // console.log(`getting comment ${id} from URL...`);

                const commentResp = await axios.get(`${process.env.HACKERNEWS_API_URL}/item/${id}.json`);
                const commentData: Record<string, any>[] = commentResp.data;

                cache.put(`comment:${id}`, commentData, 3600);
            }

            comments.push(cache.get(`comment:${id}`));
        }

        return new Promise(resolve => resolve(comments));
    };

    return commentsResolver;
}
