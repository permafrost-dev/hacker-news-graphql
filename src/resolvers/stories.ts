import axios from 'axios';

export function getResolver(topStoryIds, topStories, cache: any): any {
    const stories = async (_, { first, kind }) => {
        topStoryIds.splice(0, topStoryIds.length);
        topStories.splice(0, topStories.length);

        kind = `${kind}`.toLowerCase();

        const ids: number[] = [];
        const topstoryCacheKey = `${kind}storyids:${first}`;

        if (!cache.has(topstoryCacheKey)) {
            // console.log(`getting ${kind}storyids from URL...`);

            const resp = await axios.get(`${process.env.HACKERNEWS_API_URL}/${kind}stories.json?limitToFirst=${first}&orderBy="$key"`);
            const data: number[] = resp.data;

            cache.put(topstoryCacheKey, data, 60);
            ids.push(...data);
        }

        const cacheData = cache.get(topstoryCacheKey);

        ids.push(...cacheData);
        ids.splice(first);

        const storyDataPromises = ids
            .map(id => {
                if (cache.has(`${kind}story:${id}`)) {
                    return null;
                }

                // console.log(`getting story ${id} from URL...`);

                return axios.get(`${process.env.HACKERNEWS_API_URL}/item/${id}.json`);
            })
            .filter(item => item !== null);

        (await Promise.all(storyDataPromises))
            .map(resp => resp?.data)
            .filter(item => item !== null)
            .forEach(item => {
                cache.put(`${kind}story:${item.id}`, item, 600);
            });

        for (const id of ids) {
            topStories.push(cache.get(`${kind}story:${id}`));
        }

        return new Promise(resolve => resolve(topStories));
    };

    return stories;
}
