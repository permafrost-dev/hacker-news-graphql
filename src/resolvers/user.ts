/* eslint-disable no-unused-vars */

import { MemoryCache } from '../MemoryCache';
import axios from 'axios';

export function getResolver(cache: MemoryCache): any {
    const userResolver = async (_, { id }) => {
        if (cache.has(`user:${id}`)) {
            return cache.get(`user:${id}`);
        }

        const resp = await axios.get(`${process.env.HACKERNEWS_API_URL}/user/${id}.json`);

        cache.put(`user:${id}`, resp.data, 3600);

        return cache.get(`user:${id}`);
    };

    return userResolver;
}
