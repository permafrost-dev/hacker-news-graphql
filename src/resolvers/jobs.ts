import { Cache } from '@/lib/cache/Cache';
import axios from 'axios';
const hash = require('object-hash');

export function getResolver(storyIds, stories, cache: Cache): any {
    const result = async (_, { first, skipText }, { req }) => {
        const kind = 'job';

        storyIds.splice(0, storyIds.length);
        stories.splice(0, stories.length);

        const queryHash = hash(req.body);

        const ids: number[] = [];
        //const cacheKey = `jobstoryids:${first}-${skipText ? 'skipText' : 'dontSkipText'}`;
        const cacheKey = `jobstoryids:${first}:${queryHash}`;

        console.log('cacheKey = ', cacheKey);

        const hasKey = await cache.has(cacheKey);

        if (!hasKey) {
            console.log(`getting ${kind}storyids from URL...`);

            const resp = await axios.get(`${process.env.HACKERNEWS_API_URL}/jobstories.json?limitToFirst=${first + 5}&orderBy="$key"`);
            const data: number[] = resp.data;

            cache.put(cacheKey, data, 600);
            ids.push(...data);
        }

        const cacheData = await cache.get(cacheKey);

        ids.push(...cacheData);
        ids.splice(first);

        const storyDataPromises = ids
            .map(async id => {
                if (await cache.has(`${kind}story:${id}`)) {
                    return cache.get(`${kind}story:${id}`);
                }

                console.log(`getting story ${id} from URL...`);

                return axios.get(`${process.env.HACKERNEWS_API_URL}/item/${id}.json`);
            })
            .filter(item => item !== null);

        (await Promise.all(storyDataPromises))
            .map((resp: any) => resp?.data)
            .filter(item => item !== null)
            .filter(item => typeof item !== 'undefined')
            .forEach(item => {
                try {
                    cache.put(`${kind}story:${item.id}`, item, 3600);
                } catch (err) {
                    console.log(err);
                }
            });
        for (const id of ids) {
            const storyItem = await cache.get(`${kind}story:${id}`);

            if (!skipText || typeof storyItem['text'] === 'undefined' || storyItem['text'] === null) {
                stories.push(storyItem);
            }
        }

        return new Promise(resolve => resolve(stories));
    };

    return result;
}
