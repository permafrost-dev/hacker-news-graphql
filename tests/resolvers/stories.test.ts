/* eslint-disable no-undef */

import { MemoryCache } from '@/MemoryCache';
import { getResolver } from '@/resolvers/stories';

let cache: MemoryCache;
let resolver: any;
const storyIds: number[] = [];
const stories: Record<string, any>[] = [];

process.env.HACKERNEWS_API_URL = 'http://127.0.0.1:3031';

beforeEach(() => {
    cache = new MemoryCache();
    resolver = getResolver(storyIds, stories, cache);
});

it('it retrieves the story data and returns it', async () => {
    const data = await resolver({}, { first: 2, kind: 'NEW' });

    for (const item of data) {
        //item.ts = 0;
        item.expires = 0;
    }

    expect(data).toMatchSnapshot();
});

it('it retrieves the user data and caches it', async () => {
    await resolver({}, { first: 2, kind: 'NEW' });

    expect(cache.has('newstoryids:2')).toBeTruthy();
});

it('it retrieves each story and caches it', async () => {
    const data = await resolver({}, { first: 2, kind: 'NEW' });

    data.forEach(item => {
        expect(cache.has(`newstory:${item.id}`)).toBeTruthy();
    });

    for (const item in cache.map) {
        cache.map[item].ts = 0;
        cache.map[item].expires = 0;
    }

    expect(cache.map).toMatchSnapshot();
});
