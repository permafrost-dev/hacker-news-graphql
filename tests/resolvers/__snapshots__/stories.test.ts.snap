// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it retrieves each story and caches it 1`] = `
Object {
  "newstory:27397208": Object {
    "expires": 0,
    "ts": 0,
    "value": Object {
      "by": "brundolf",
      "descendants": 35,
      "id": 27397208,
      "kids": Array [
        27397733,
        27397758,
        27398262,
        27397815,
        27398632,
        27398080,
        27397746,
      ],
      "score": 331,
      "time": 1622831487,
      "title": "Pokegb: A gameboy emulator that only plays Pokémon Blue, in 68 lines of C++",
      "type": "story",
      "url": "https://binji.github.io/posts/pokegb/",
    },
  },
  "newstory:27399416": Object {
    "expires": 0,
    "ts": 0,
    "value": Object {
      "by": "tosh",
      "descendants": 65,
      "id": 27399416,
      "kids": Array [
        27400240,
        27399935,
        27399745,
        27399824,
        27399455,
        27399804,
        27399957,
        27399894,
        27399952,
        27399844,
        27399618,
        27399809,
        27399554,
        27400151,
      ],
      "score": 110,
      "time": 1622843033,
      "title": "John Carmack on JPEG",
      "type": "story",
      "url": "https://twitter.com/ID_AA_Carmack/status/1400930510671601666",
    },
  },
  "newstoryids:2": Object {
    "expires": 0,
    "ts": 0,
    "value": Array [
      27399416,
      27397208,
    ],
  },
}
`;

exports[`it retrieves the story data and returns it 1`] = `
Array [
  Object {
    "by": "tosh",
    "descendants": 65,
    "expires": 0,
    "id": 27399416,
    "kids": Array [
      27400240,
      27399935,
      27399745,
      27399824,
      27399455,
      27399804,
      27399957,
      27399894,
      27399952,
      27399844,
      27399618,
      27399809,
      27399554,
      27400151,
    ],
    "score": 110,
    "time": 1622843033,
    "title": "John Carmack on JPEG",
    "type": "story",
    "url": "https://twitter.com/ID_AA_Carmack/status/1400930510671601666",
  },
  Object {
    "by": "brundolf",
    "descendants": 35,
    "expires": 0,
    "id": 27397208,
    "kids": Array [
      27397733,
      27397758,
      27398262,
      27397815,
      27398632,
      27398080,
      27397746,
    ],
    "score": 331,
    "time": 1622831487,
    "title": "Pokegb: A gameboy emulator that only plays Pokémon Blue, in 68 lines of C++",
    "type": "story",
    "url": "https://binji.github.io/posts/pokegb/",
  },
]
`;
